hash_practice/exercises.py

```diff
@@ -1,19 +1,73 @@
 
+from audioop import reverse
+from copy import copy
+from curses import keyname
+from tabnanny import check
+
+
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n^2)
+        Space Complexity: O(n)
     """
-    pass
+    if not strings:
+        return strings
+
+    lists_of_anagrams = []
+    list_of_dicts = []
+
+    for string in strings:
+        word_dict = {}
+        for letter in string:
+            if letter in word_dict:
+                word_dict[letter] += 1
+            else:
+                word_dict[letter] = 1
+        if word_dict not in list_of_dicts:
+            list_of_dicts.append(word_dict)
+            lists_of_anagrams.append([string])
+        else:
+            index = list_of_dicts.index(word_dict)
+            lists_of_anagrams[index] += [string]
+
+    return lists_of_anagrams
+
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n log(n))
+        Space Complexity: O(n)
     """
-    pass
+    if not nums:
+        return nums
+    
+    freq_map = {}
+    for num in nums:
+        if num in freq_map:
+            freq_map[num] += 1
+        else:
+            freq_map[num] = 1
+
+    if len(freq_map) == k:
+        return nums
+
+    holding = []
+
+    for key, value in freq_map.items():
+        holding.append((key, value))
+    
+    holding.sort(key = lambda x: x[1], reverse=True)
+    
+    result = []
+
+    for tuple in holding:
+        result.append(tuple[0]) 
+
+    return result[:k]
+    
+
 
 
 def valid_sudoku(table):

```

Comment:
✨ Nice! We can actually say time and space complexity are O(1) since the grid will always be a fixed size: 9 x 9.